pkg/agent/storage/utils.go

```diff
@@ -56,3 +67,58 @@ func RemoveDir(dir string) error {
 	}
 	return nil
 }
+
+func GetProvider(providers map[Protocol]Provider, protocol Protocol) (Provider, error) {
+	if provider, ok := providers[protocol]; ok {
+		return provider, nil
+	}
+
+	switch protocol {
+	case GCS:
+		var gcsClient *gstorage.Client
+		var err error
+
+		ctx := context.Background()
+		if _, ok := os.LookupEnv(gcscredential.GCSCredentialEnvKey); ok {
+			// GCS relies on environment variable GOOGLE_APPLICATION_CREDENTIALS to point to the service-account-key
+			// If set, it will be automatically be picked up by the client.
+			gcsClient, err = gstorage.NewClient(ctx)
+		} else {
+			gcsClient, err = gstorage.NewClient(ctx, option.WithoutAuthentication())
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		providers[GCS] = &GCSProvider{
+			Client: stiface.AdaptClient(gcsClient),
+		}
+	case S3:
+		if endpoint, ok := os.LookupEnv(s3credential.AWSEndpointUrl); ok {
```

Comment:
If we are doing public gcs, can we also do public s3?

Reply:
The AWS SDK for Go requires credentials so we can't do the same thing here

Reply:
@PatrickXYS can you help review if we are doing correctly?